File: HotDesk_SPICE/App_Module/Search.ascx.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Web.UI;

/// <summary>
/// Search User Control
/// Handler Basic Search and Advanced Search
/// </summary>
public partial class Search : System.Web.UI.UserControl
{
    private List<string> _Search;
    private List<string> _query;

    protected void Page_Load(object sender, EventArgs e)
    {
        if (!IsPostBack)
        {
            BindDDL();
        }

        Control.Base setting = (Control.Base)this.Page;

        if (Session["key"] == null)
        {
            Session["key"] = setting.SetupKey;
        }
        else
        {
            if (Session["key"].ToString() != setting.SetupKey)
            {
                Session["key"] = setting.SetupKey;
                Session["Type"] = null;
                flush();
            }
        }

        if (Session["Type"] == null)
        {
            ChangeType("B");
        }
        else
        {
            ChangeType(Session["Type"].ToString());
            _Search = (List<string>)Session["Search"];
            _query = (List<string>)Session["Query"];
        }

        if (!IsPostBack)
        {
            if (Session["Type"] != null && Session["Type"].ToString() == "B")
            {
                if (setting.SearchField.Trim() != string.Empty)
                {
                    txtSearch.Text = setting.SearchValue;
                    ddlSearch.SelectedValue = setting.SearchField;
                }
            }
            else if (Session["Search"] != null)
            {
                BindCreteria((List<string>)Session["Search"]);
            }
        }

        btnReset.Enabled = searchCriteria.Items.Count > 0;
        btnMinus.Enabled = searchCriteria.Items.Count > 0;
    }

    protected void BindDDL()
    {
        Control.Base setting = (Control.Base)this.Page;

        chkDeleted.Visible = setting.DeleteControl;

        if (setting.DeleteControl)
        {
            chkDeleted.Visible = setting.ShowDeletedControl;
            chkDeleted.Checked = setting.ShowDeleted;
        }

        if (setting.SetupKey == string.Empty)
        {
            this.Visible = false;
        }
        else
        {
        Library.Root.Control.Binding.BindDropDownListResource(ddlSearch, setting.SetupKey, "-", "-");
        Library.Root.Control.Binding.BindDropDownListResource(ddlSearchUsing, setting.SetupKey, "-", "-");
        }
    }

    private void ChangeType(string Type)
    {
        Session["Type"] = Type;

        if (Type == "A")
        {
            lblSearch.Text = "Advance Search";
            pnlBasic.Visible = false;
            pnlAdvance.Visible = true;
        }
        else if (Type == "B")
        {
            lblSearch.Text = "Basic Search";
            pnlBasic.Visible = true;
            pnlAdvance.Visible = false;
        }
    }

    private void detector()
    {
        ddlSearch2.Enabled = _Search.Count > 0;
    }

    private void resetController()
    {
        ddlSearch.SelectedIndex = 0;
        ddlSearchUsing.SelectedIndex = 0;
        ddlOperator1.SelectedIndex = 0;
        ddlSearch2.SelectedIndex = 0;
        txtSearch.Text = string.Empty;
        txtSearchUsing.Text = string.Empty;
    }

    private void flush()
    {
        _Search = new List<string>();
        _query = new List<string>();
        resetController();
        searchCriteria.Items.Clear();
        Session["Search"] = _Search;
        Session["Query"] = _query;
    }

    /// <summary>
    /// Retrieve the delete value from the checkbox
    /// </summary>
    public bool Deleted
    {
        get { return chkDeleted.Checked; }
    }

    public string SearchType
    {
        get { return Session["Type"] != null ? Session["Type"].ToString() : string.Empty; }
    }

    protected void lbAdvSearch_Click(object sender, EventArgs e)
    {
        flush();
        ChangeType("A");
    }

    protected void lbBasicSearch_Click(object sender, EventArgs e)
    {
        flush();
        ChangeType("B");
    }

    protected void btnSubmit_Click(object sender, EventArgs e)
    {
        Control.Base setting = (Control.Base)this.Page;

        if (ddlSearch.SelectedIndex > 0)
        {
            setting.SearchField = ddlSearch.SelectedValue;
            setting.SearchValue = txtSearch.Text;
        }
        else
        {
            setting.SearchField = string.Empty;
            setting.SearchValue = string.Empty;
        }

        setting.PageNo = 1;
        Response.Redirect(setting.GenerateList);
    }

    protected void btnSubmit2_Click(object sender, EventArgs e)
    {
        Control.Base setting = (Control.Base)this.Page;
        setting.SearchField = string.Empty;
        setting.SearchValue = ConvertionString(_query);

        setting.PageNo = 1;
        Response.Redirect(setting.GenerateList);
    }

    protected void btnPlus_Click(object sender, EventArgs e)
    {
        if (!CheckSelection(ddlSearchUsing.SelectedValue, "Search Field")) return;
        if (!CheckSelection(ddlOperator1.SelectedValue, "Operator")) return;

        if (_Search.Count > 0)
        {
            if (!CheckSelection(ddlSearch2.SelectedValue, "Operator (And/Or)")) return;
        }

        if (!EmptyCheck(txtSearchUsing.Text, "Search Value")) return;

        _Search.Add((_Search.Count > 0 ? ddlSearch2.SelectedItem.Text.Replace("-", "") + " " : string.Empty)
            + ddlSearchUsing.SelectedItem.Text + " " + ddlOperator1.SelectedItem.Text + " " + txtSearchUsing.Text);
        AddQuery(
            _Search.Count > 0 ? ddlSearch2.SelectedItem.Text.Replace("-", "") + " " : "Add ",
            ddlSearchUsing.SelectedValue,
            ddlOperator1.SelectedValue,
            txtSearchUsing.Text);

        Session["Search"] = _Search;
        Session["Query"] = _query;

        BindCreteria(_Search);
        detector();
        resetController();
        ddlSearchUsing.Focus();
    }

    private void AddQuery(string Addtional, string SearchField, string Operator, string SearchValue)
    {
        if (Operator.Trim().ToUpper() == "LIKE")
            SearchValue = "'%" + SearchValue + "%'";
        else
            SearchValue = "'" + SearchValue + "'";

        _query.Add(Addtional + " UPPER(" + SearchField + ") " + Operator + " UPPER(" + SearchValue + ")");
    }

    private bool CheckSelection(string Value, string field)
    {
        if (Value == "-")
        {
            Control.MessageCenter.ShowAJAXMessageBox(this.Page, field + " was " + Resources.Message.InvalidSelect);
            return false;
        }
        return true;
    }

    private bool EmptyCheck(string Value, string Field)
    {
        if (Value.Trim() == string.Empty)
        {
            Control.MessageCenter.ShowAJAXMessageBox(this.Page, string.Format(Resources.Message.FieldEmpty, Field));
            return false;
        }
        return true;
    }

    private string ConvertionString(List<string> list)
    {
        var temp = new StringBuilder();
        foreach (string str in list)
            temp.AppendLine(str);
        return temp.ToString();
    }

    private void BindCreteria(List<string> list)
    {
        btnReset.Enabled = list.Count > 0;
        btnMinus.Enabled = list.Count > 0;
        searchCriteria.DataSource = list;
        searchCriteria.DataBind();
    }

    protected void btnMinus_Click(object sender, EventArgs e)
    {
        if (_Search.Count == 0) return;

        if (searchCriteria.SelectedIndex == -1)
        {
            if (_Search.Count == 1)
            {
                _Search = new List<string>();
                _query = new List<string>();
            }
            else
            {
                _Search.RemoveAt(_Search.Count - 1);
                _query.RemoveAt(_query.Count - 1);
            }
        }
        else
        {
            _Search.RemoveAt(searchCriteria.SelectedIndex);
            _query.RemoveAt(searchCriteria.SelectedIndex);
        }

        if (_Search.Count > 0)
        {
            if (_Search[0].Substring(0, 3) == "AND")
                _Search[0] = _Search[0].Substring(3);
            else if (_Search[0].Substring(0, 2) == "OR")
                _Search[0] = _Search[0].Substring(2);
        }

        if (_query.Count > 0)
        {
            if (_query[0].Substring(0, 3) == "AND")
                _query[0] = _query[0].Substring(3);
            else if (_query[0].Substring(0, 2) == "OR")
                _query[0] = _query[0].Substring(2);

            _query[0] = "AND" + _query[0];
        }

        Session["Search"] = _Search;
        Session["Query"] = _query;

        BindCreteria(_Search);
        detector();
        resetController();
        ddlSearchUsing.Focus();
    }

    protected void btnReset_Click(object sender, EventArgs e)
    {
        _Search = new List<string>();
        _query = new List<string>();
        Session["Search"] = _Search;
        Session["Query"] = _query;

        BindCreteria(_Search);
        detector();
        resetController();
    }

    protected void chkDeleted_CheckedChanged(object sender, EventArgs e)
    {
        Control.Base setting = (Control.Base)this.Page;
        setting.ShowDeleted = chkDeleted.Checked;
        Response.Redirect(setting.GenerateList);
    }
}

File: TMS_HotDesk/App_Module/Search.ascx.vb
''' <summary>
''' Search User Control 
''' Handler Basic Search and Advanced Search 
''' 
''' Remark : Based on previous Version and modified the way of the binding
''' ----------------------------------------
''' C.C.Yeon    25 APril 2011  Modified 
''' </summary>
''' <remarks></remarks>

Partial Class Search
    Inherits System.Web.UI.UserControl

    Private Search As Collections.Generic.List(Of String)
    Private query As Collections.Generic.List(Of String)

    Protected Sub Page_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Load
        If Not IsPostBack Then
            Me.BindDDL()
        End If

        Dim setting As Control.Base = CType(Me.Page, Control.Base)

        If Session("key") Is Nothing Then
            Session("key") = setting.SetupKey
        Else
            If Session("key") <> setting.SetupKey Then
                Session("key") = setting.SetupKey
                Session("Type") = Nothing
                flush()
            End If
        End If

        If Session("Type") Is Nothing Then
            ChangeType("B")
        Else
            ChangeType(Session("Type"))
            Search = Session("Search")
            query = Session("Query")
        End If

        If Not IsPostBack Then
            If Session("Type") = "B" Then
                If setting.SearchField.Trim <> String.Empty Then
                    txtSearch.Text = setting.SearchValue
                    ddlSearch.SelectedValue = setting.SearchField
                End If
            Else
                BindCreteria(Session("Search"))
            End If
        End If

        btnReset.Enabled = searchCriteria.Items.Count > 0
        btnMinus.Enabled = searchCriteria.Items.Count > 0
    End Sub

    Protected Sub BindDDL()
        'Retrieve the value of the properties from the page
        'Based on the key retrieve the value from the resource and bind its into search selection
        Dim setting As Control.Base = CType(Me.Page, Control.Base)

        'here to disappear the delete check box
        Me.chkDeleted.Visible = setting.DeleteControl

        If setting.DeleteControl Then
            Me.chkDeleted.Visible = setting.ShowDeletedControl
            Me.chkDeleted.Checked = setting.ShowDeleted
        End If

        If setting.SetupKey = String.Empty Then
            Me.Visible = False
        Else
            Control.Binding.BindDropDownListResource(Me.ddlSearch, setting.SetupKey, "-", "-")
            Control.Binding.BindDropDownListResource(Me.ddlSearchUsing, setting.SetupKey, "-", "-")
        End If
    End Sub

    Private Sub ChangeType(ByVal Type As String)
        Session("Type") = Type

        If Type = "A" Then
            lblSearch.Text = "Advance Search"
            pnlBasic.Visible = False
            pnlAdvance.Visible = True
        ElseIf Type = "B" Then
            lblSearch.Text = "Basic Search"
            pnlBasic.Visible = True
            pnlAdvance.Visible = False
        End If
    End Sub

    Private Sub detector()
        If Search.Count = 0 Then
            ddlSearch2.Enabled = False
        Else
            ddlSearch2.Enabled = True
        End If
    End Sub

    Private Sub resetController()
        ddlSearch.SelectedIndex = 0
        ddlSearchUsing.SelectedIndex = 0
        ddlOperator1.SelectedIndex = 0
        ddlSearch2.SelectedIndex = 0
        txtSearch.Text = String.Empty
        txtSearchUsing.Text = String.Empty
    End Sub

    Private Sub flush()
        Search = New Collections.Generic.List(Of String)
        query = New Collections.Generic.List(Of String)
        resetController()
        searchCriteria.Items.Clear()
        Session("Search") = Search
        Session("Query") = query
    End Sub

    ''' <summary>
    ''' Retrieve the delete value from the checkedbox
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property Deleted() As Boolean
        Get
            Return chkDeleted.Checked
        End Get
    End Property

    Public ReadOnly Property SearchType() As String
        Get
            Return Session("Type")
        End Get
    End Property

    Protected Sub lbAdvSearch_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles lbAdvSearch.Click
        flush()
        ChangeType("A")
    End Sub

    Protected Sub lbBasicSearch_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles lbBasicSearch.Click
        flush()
        ChangeType("B")
    End Sub

    Protected Sub btnSubmit_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnSubmit.Click
        Dim setting As Control.Base = CType(Me.Page, Control.Base)

        If ddlSearch.SelectedIndex > 0 Then
            setting.SearchField = ddlSearch.SelectedValue
            setting.SearchValue = txtSearch.Text
        Else
            setting.SearchField = String.Empty
            setting.SearchValue = String.Empty
        End If

        setting.PageNo = 1
        Response.Redirect(setting.GenerateList)
    End Sub

    Protected Sub btnSubmit2_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnSubmit2.Click
        Dim setting As Control.Base = CType(Me.Page, Control.Base)
        setting.SearchField = String.Empty
        setting.SearchValue = ConvertionString(query)

        setting.PageNo = 1
        Response.Redirect(setting.GenerateList)
    End Sub

    Protected Sub btnPlus_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnPlus.Click
        If CheckSelection(ddlSearchUsing.SelectedValue, "Search Field") = False Then
            Return
        End If

        If CheckSelection(ddlOperator1.SelectedValue, "Operator") = False Then
            Return
        End If

        If Search.Count > 0 Then
            If CheckSelection(ddlSearch2.SelectedValue, "Operator (And/Or)") = False Then
                Return
            End If
        End If

        If EmptyCheck(txtSearchUsing.Text, "Search Value") = False Then
            Return
        End If

        Search.Add(If(Search.Count > 0, ddlSearch2.SelectedItem.Text.Replace("-", "") & " ", String.Empty) & ddlSearchUsing.SelectedItem.Text & " " & ddlOperator1.SelectedItem.Text & " " & txtSearchUsing.Text)
        AddQuery(If(Search.Count > 0, ddlSearch2.SelectedItem.Text.Replace("-", "") & " ", "Add "), ddlSearchUsing.SelectedValue, ddlOperator1.SelectedValue, txtSearchUsing.Text)

        Session("Search") = Search
        Session("Query") = query

        BindCreteria(Search)
        detector()
        resetController()
        ddlSearchUsing.Focus()
    End Sub

    Private Sub AddQuery(ByVal Addtional As String, ByVal SearchField As String, ByVal [Operator] As String, ByVal SearchValue As String)
        If [Operator].Trim.ToUpper = "LIKE" Then
            SearchValue = "'%" & SearchValue & "%'"
        Else
            SearchValue = "'" & SearchValue & "'"
        End If

        query.Add(Addtional & " UPPER(" & SearchField & ") " & [Operator] & " UPPER(" & SearchValue & ")")
    End Sub

    Private Function CheckSelection(ByVal Value As String, ByVal field As String) As Boolean
        If Value = "-" Then
            Library.Root.Control.MessageCenter.ShowAJAXMessageBox(Me.Page, field & " was " & Resources.Message.InvalidSelect)
            Return False
        Else
            Return True
        End If
    End Function

    Private Function EmptyCheck(ByVal Value As String, ByVal Field As String) As Boolean
        If Value.Trim = String.Empty Then
            Library.Root.Control.MessageCenter.ShowAJAXMessageBox(Me.Page, String.Format(Resources.Message.FieldEmpty, Field))
            Return False
        Else
            Return True
        End If
    End Function

    Private Function ConvertionString(ByVal list As Collections.Generic.List(Of String)) As String
        Dim temp As New Text.StringBuilder
        For Each Str As String In list
            temp.AppendLine(Str)
        Next

        Return temp.ToString
    End Function

    Private Sub BindCreteria(ByVal list As Collections.Generic.List(Of String))
        btnReset.Enabled = list.Count > 0
        btnMinus.Enabled = list.Count > 0

        searchCriteria.DataSource = list
        searchCriteria.DataBind()
    End Sub

    Protected Sub btnMinus_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnMinus.Click
        If Search.Count = 0 Then
            Exit Sub
        End If

        If searchCriteria.SelectedIndex = -1 Then
            If Search.Count = 1 Then
                Search = New Collections.Generic.List(Of String)
                query = New Collections.Generic.List(Of String)
            Else
                Search.RemoveAt(Search.Count - 1)
                query.RemoveAt(Search.Count - 1)
            End If
        Else
            Search.RemoveAt(searchCriteria.SelectedIndex)
            query.RemoveAt(searchCriteria.SelectedIndex)
        End If

        If Search.Count > 0 Then
            If Search.Item(0).Substring(0, 3) = "AND" Then
                Search.Item(0) = Search.Item(0).Substring(3, Search.Item(0).Length - 3)
            ElseIf Search.Item(0).Substring(0, 2) = "OR" Then
                Search.Item(0) = Search.Item(0).Substring(2, Search.Item(0).Length - 2)
            End If
        End If

        If query.Count > 0 Then
            If query.Item(0).Substring(0, 3) = "AND" Then
                query.Item(0) = query.Item(0).Substring(3, query.Item(0).Length - 3)
            ElseIf query.Item(0).Substring(0, 2) = "OR" Then
                query.Item(0) = query.Item(0).Substring(2, query.Item(0).Length - 2)
            End If

            query.Item(0) = "AND" & query.Item(0)
        End If

        Session("Search") = Search
        Session("Query") = query

        BindCreteria(Search)
        detector()
        resetController()
        ddlSearchUsing.Focus()
    End Sub

    Protected Sub btnReset_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnReset.Click
        Search = New Collections.Generic.List(Of String)
        query = New Collections.Generic.List(Of String)
        Session("Search") = Search
        Session("Query") = query

        BindCreteria(Search)
        detector()
        resetController()
    End Sub

    Protected Sub chkDeleted_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles chkDeleted.CheckedChanged
        Dim setting As Control.Base = CType(Me.Page, Control.Base)
        setting.ShowDeleted = chkDeleted.Checked
        Response.Redirect(setting.GenerateList)
    End Sub

End Class
